src/main/host/descriptor/pipe.rs

```diff
@@ -302,10 +301,13 @@ pub struct SharedBuf {
 }
 
 impl SharedBuf {
-    pub fn new() -> Self {
+    pub fn new(max_len: usize) -> Self {
+        assert_ne!(max_len, 0);
         Self {
-            queue: ByteQueue::new(8192),
-            max_len: c::CONFIG_PIPE_BUFFER_SIZE as usize,
+            // 4048 bytes hopefully leaves enough space (4096-4048 = 48 bytes) to fit each chunk
+            // within a single page
+            queue: ByteQueue::new(4048),
```

Comment:
The 48 and 4048 numbers here are arbitrary since I don't have any insight into how much memory overhead the `Bytes` structure takes.

Reply:
The value passed to `ByteQueue::new` is the number of bytes that will be pointed to by a `Bytes` or `BytesMut` object. But we don't really know how those bytes are allocated. I was worried that maybe `Bytes` would store some data in the same allocation, so for example if we wanted to allocate 4096 bytes, maybe `Bytes` would prepend 8 bytes and perform an allocation of 5004 bytes, which would then be spread across 2 pages instead of one. The `Bytes` structure itself is very small (a pointer, a length, and some vtable data).

The byte queue uses `BytesMut::from_iter(std::iter::repeat(0).take(size))`. It looks like `BytesMut::from_iter` uses `BytesMut::from_vec(Vec::from_iter(into_iter))`, and at higher optimization levels, it looks like `Vec::from_iter()` probably does just an `alloc()` and then a `memset()`, and I don't think a `Vec` stores any other data in the same allocation, so maybe it's fine to just pass 4096 to `ByteQueue::new`.

Reply:
I made this 4096 bytes now instead, since I think the final alloc call will request the same number of bytes.